### PR TITLE
Improved typings.

### LINKED ISSUE #904: Ts error: Unused '@ts-expect-error' directive.  TS2578
I'm, using typescript 4.2.2
uniforms 3.2.1

/node_modules/uniforms-bootstrap4/src/gridClassName.ts
TypeScript error in /node_modules/uniforms-bootstrap4/src/gridClassName.ts(47,9):
Unused '@ts-expect-error' directive.  TS2578

    45 |       (Object.keys(grid) as GridSize[])
    46 |         .sort(compareSizeClass)
  > 47 |         // @ts-expect-error Weird type refinement problem.
       |         ^
    48 |         .map(size => gridClassNamePart(size, grid[size], side))
    49 |         .join(' ')
    50 |     );

### PR DESCRIPTION
**Pull request by radekmie**
This PR improves typings in a few ways:
* Update to TypeScript 4.2.4 makes some of the generated `.d.ts` files much shorter.
    * This _may_ have fixed #904.
    * This required `tslib` update, thus this PR supersedes #948.
* Overhaul of the typings in `createAutoField` and `connectField`.
    * The former is trivial: extracting "temporary" types into named and exported ones makes them usable in the `.d.ts` files.
    * The latter was similar, however, these types were more complex. As a result, there are three new types: `ConnectFieldOptions`, `ConnectedFieldProps<Props, Value>`, and `ConnectedField<Props, Value`.
        * All three got marked as `@internal` and I think that's the way to go - let's mark "actually internal" types with this JSDoc.

As a result, multiple `.d.ts` files are **much** smaller now.

<details>
<summary><code>uniforms-antd/es5/AutoField.d.ts</code> before - 2831B</summary>

```ts
/// <reference types="react" />
export declare type AutoFieldProps = Parameters<typeof AutoField>[0];
declare const AutoField: {
    (rawProps: import("uniforms").Override<Record<string, unknown>, {
        component?: import("react").ComponentClass<any, any> | import("react").FunctionComponent<any> | (import("react").FunctionComponent<import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<any>>, import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<unknown>>, {
            label?: import("react").ReactNode;
            name: string;
            placeholder?: string | boolean | null | undefined;
        }>>> & {
            Component: import("react").ComponentType<Partial<import("uniforms").GuaranteedProps<any>>>;
            options: {
                initialValue?: boolean | undefined;
                kind?: "leaf" | "node" | undefined;
            } | undefined;
        }) | undefined;
        name: string;
    }>): import("react").ReactElement<any, string | ((props: any) => import("react").ReactElement<any, any> | null) | (new (props: any) => import("react").Component<any, any, any>)>;
    componentDetectorContext: import("react").Context<(props: import("uniforms").GuaranteedProps<unknown> & Record<string, any>, uniforms: import("uniforms").Context<unknown>) => import("react").ComponentClass<any, any> | import("react").FunctionComponent<any> | (import("react").FunctionComponent<import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<any>>, import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<unknown>>, {
        label?: import("react").ReactNode;
        name: string;
        placeholder?: string | boolean | null | undefined;
    }>>> & {
        Component: import("react").ComponentType<Partial<import("uniforms").GuaranteedProps<any>>>;
        options: {
            initialValue?: boolean | undefined;
            kind?: "leaf" | "node" | undefined;
        } | undefined;
    })>;
    defaultComponentDetector: (props: import("uniforms").GuaranteedProps<unknown> & Record<string, any>, uniforms: import("uniforms").Context<unknown>) => import("react").ComponentClass<any, any> | import("react").FunctionComponent<any> | (import("react").FunctionComponent<import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<any>>, import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<unknown>>, {
        label?: import("react").ReactNode;
        name: string;
        placeholder?: string | boolean | null | undefined;
    }>>> & {
        Component: import("react").ComponentType<Partial<import("uniforms").GuaranteedProps<any>>>;
        options: {
            initialValue?: boolean | undefined;
            kind?: "leaf" | "node" | undefined;
        } | undefined;
    });
};
export default AutoField;
```

</details>

<details>
<summary><code>uniforms-antd/es5/AutoField.d.ts</code> after - 432B</summary>

```ts
/// <reference types="react" />
export { AutoFieldProps } from 'uniforms';
declare const AutoField: ((rawProps: import("uniforms").AutoFieldProps) => import("react").ReactElement<any, string | import("react").JSXElementConstructor<any>>) & {
    componentDetectorContext: import("react").Context<import("uniforms").ComponentDetector>;
    defaultComponentDetector: import("uniforms").ComponentDetector;
};
export default AutoField;
```

</details>

<details>
<summary><code>uniforms-antd/es5/TextField.d.ts</code> before - 11238B</summary>

```ts
import Input, { InputProps } from 'antd/lib/input';
import React, { Ref } from 'react';
import { FieldProps } from 'uniforms';
export declare type TextFieldProps = FieldProps<string, Omit<InputProps, 'onReset'>, {
    inputRef?: Ref<Input>;
}>;
declare const _default: React.FunctionComponent<import("uniforms").Override<import("uniforms").Override<Pick<InputProps, "hidden" | "color" | "size" | "style" | "multiple" | "disabled" | "height" | "translate" | "width" | "prefix" | "suffix" | "src" | "form" | "slot" | "title" | "pattern" | "children" | "list" | "step" | "id" | "name" | "onChange" | "placeholder" | "readOnly" | "value" | "accept" | "alt" | "autoComplete" | "autoFocus" | "capture" | "checked" | "crossOrigin" | "formAction" | "formEncType" | "formMethod" | "formNoValidate" | "formTarget" | "max" | "maxLength" | "min" | "minLength" | "required" | "type" | "defaultChecked" | "defaultValue" | "suppressContentEditableWarning" | "suppressHydrationWarning" | "accessKey" | "className" | "contentEditable" | "contextMenu" | "dir" | "draggable" | "lang" | "spellCheck" | "tabIndex" | "radioGroup" | "role" | "about" | "datatype" | "inlist" | "property" | "resource" | "typeof" | "vocab" | "autoCapitalize" | "autoCorrect" | "autoSave" | "itemProp" | "itemScope" | "itemType" | "itemID" | "itemRef" | "results" | "security" | "unselectable" | "inputMode" | "is" | "aria-activedescendant" | "aria-atomic" | "aria-autocomplete" | "aria-busy" | "aria-checked" | "aria-colcount" | "aria-colindex" | "aria-colspan" | "aria-controls" | "aria-current" | "aria-describedby" | "aria-details" | "aria-disabled" | "aria-dropeffect" | "aria-errormessage" | "aria-expanded" | "aria-flowto" | "aria-grabbed" | "aria-haspopup" | "aria-hidden" | "aria-invalid" | "aria-keyshortcuts" | "aria-label" | "aria-labelledby" | "aria-level" | "aria-live" | "aria-modal" | "aria-multiline" | "aria-multiselectable" | "aria-orientation" | "aria-owns" | "aria-placeholder" | "aria-posinset" | "aria-pressed" | "aria-readonly" | "aria-relevant" | "aria-required" | "aria-roledescription" | "aria-rowcount" | "aria-rowindex" | "aria-rowspan" | "aria-selected" | "aria-setsize" | "aria-sort" | "aria-valuemax" | "aria-valuemin" | "aria-valuenow" | "aria-valuetext" | "dangerouslySetInnerHTML" | "onCopy" | "onCopyCapture" | "onCut" | "onCutCapture" | "onPaste" | "onPasteCapture" | "onCompositionEnd" | "onCompositionEndCapture" | "onCompositionStart" | "onCompositionStartCapture" | "onCompositionUpdate" | "onCompositionUpdateCapture" | "onFocus" | "onFocusCapture" | "onBlur" | "onBlurCapture" | "onChangeCapture" | "onBeforeInput" | "onBeforeInputCapture" | "onInput" | "onInputCapture" | "onResetCapture" | "onSubmit" | "onSubmitCapture" | "onInvalid" | "onInvalidCapture" | "onLoad" | "onLoadCapture" | "onError" | "onErrorCapture" | "onKeyDown" | "onKeyDownCapture" | "onKeyPress" | "onKeyPressCapture" | "onKeyUp" | "onKeyUpCapture" | "onAbort" | "onAbortCapture" | "onCanPlay" | "onCanPlayCapture" | "onCanPlayThrough" | "onCanPlayThroughCapture" | "onDurationChange" | "onDurationChangeCapture" | "onEmptied" | "onEmptiedCapture" | "onEncrypted" | "onEncryptedCapture" | "onEnded" | "onEndedCapture" | "onLoadedData" | "onLoadedDataCapture" | "onLoadedMetadata" | "onLoadedMetadataCapture" | "onLoadStart" | "onLoadStartCapture" | "onPause" | "onPauseCapture" | "onPlay" | "onPlayCapture" | "onPlaying" | "onPlayingCapture" | "onProgress" | "onProgressCapture" | "onRateChange" | "onRateChangeCapture" | "onSeeked" | "onSeekedCapture" | "onSeeking" | "onSeekingCapture" | "onStalled" | "onStalledCapture" | "onSuspend" | "onSuspendCapture" | "onTimeUpdate" | "onTimeUpdateCapture" | "onVolumeChange" | "onVolumeChangeCapture" | "onWaiting" | "onWaitingCapture" | "onAuxClick" | "onAuxClickCapture" | "onClick" | "onClickCapture" | "onContextMenu" | "onContextMenuCapture" | "onDoubleClick" | "onDoubleClickCapture" | "onDrag" | "onDragCapture" | "onDragEnd" | "onDragEndCapture" | "onDragEnter" | "onDragEnterCapture" | "onDragExit" | "onDragExitCapture" | "onDragLeave" | "onDragLeaveCapture" | "onDragOver" | "onDragOverCapture" | "onDragStart" | "onDragStartCapture" | "onDrop" | "onDropCapture" | "onMouseDown" | "onMouseDownCapture" | "onMouseEnter" | "onMouseLeave" | "onMouseMove" | "onMouseMoveCapture" | "onMouseOut" | "onMouseOutCapture" | "onMouseOver" | "onMouseOverCapture" | "onMouseUp" | "onMouseUpCapture" | "onSelect" | "onSelectCapture" | "onTouchCancel" | "onTouchCancelCapture" | "onTouchEnd" | "onTouchEndCapture" | "onTouchMove" | "onTouchMoveCapture" | "onTouchStart" | "onTouchStartCapture" | "onPointerDown" | "onPointerDownCapture" | "onPointerMove" | "onPointerMoveCapture" | "onPointerUp" | "onPointerUpCapture" | "onPointerCancel" | "onPointerCancelCapture" | "onPointerEnter" | "onPointerEnterCapture" | "onPointerLeave" | "onPointerLeaveCapture" | "onPointerOver" | "onPointerOverCapture" | "onPointerOut" | "onPointerOutCapture" | "onGotPointerCapture" | "onGotPointerCaptureCapture" | "onLostPointerCapture" | "onLostPointerCaptureCapture" | "onScroll" | "onScrollCapture" | "onWheel" | "onWheelCapture" | "onAnimationStart" | "onAnimationStartCapture" | "onAnimationEnd" | "onAnimationEndCapture" | "onAnimationIteration" | "onAnimationIterationCapture" | "onTransitionEnd" | "onTransitionEndCapture" | "prefixCls" | "allowClear" | "bordered" | "enterKeyHint" | "onPressEnter" | "addonBefore" | "addonAfter">, import("uniforms").GuaranteedProps<string> & {
    inputRef?: ((instance: Input | null) => void) | React.RefObject<Input> | null | undefined;
}>, import("uniforms").Override<Partial<import("uniforms").GuaranteedProps<string | undefined>>, {
    label?: React.ReactNode;
    name: string;
    placeholder?: string | boolean | null | undefined;
}>>> & {
    Component: React.ComponentType<import("uniforms").Override<Pick<InputProps, "hidden" | "color" | "size" | "style" | "multiple" | "disabled" | "height" | "translate" | "width" | "prefix" | "suffix" | "src" | "form" | "slot" | "title" | "pattern" | "children" | "list" | "step" | "id" | "name" | "onChange" | "placeholder" | "readOnly" | "value" | "accept" | "alt" | "autoComplete" | "autoFocus" | "capture" | "checked" | "crossOrigin" | "formAction" | "formEncType" | "formMethod" | "formNoValidate" | "formTarget" | "max" | "maxLength" | "min" | "minLength" | "required" | "type" | "defaultChecked" | "defaultValue" | "suppressContentEditableWarning" | "suppressHydrationWarning" | "accessKey" | "className" | "contentEditable" | "contextMenu" | "dir" | "draggable" | "lang" | "spellCheck" | "tabIndex" | "radioGroup" | "role" | "about" | "datatype" | "inlist" | "property" | "resource" | "typeof" | "vocab" | "autoCapitalize" | "autoCorrect" | "autoSave" | "itemProp" | "itemScope" | "itemType" | "itemID" | "itemRef" | "results" | "security" | "unselectable" | "inputMode" | "is" | "aria-activedescendant" | "aria-atomic" | "aria-autocomplete" | "aria-busy" | "aria-checked" | "aria-colcount" | "aria-colindex" | "aria-colspan" | "aria-controls" | "aria-current" | "aria-describedby" | "aria-details" | "aria-disabled" | "aria-dropeffect" | "aria-errormessage" | "aria-expanded" | "aria-flowto" | "aria-grabbed" | "aria-haspopup" | "aria-hidden" | "aria-invalid" | "aria-keyshortcuts" | "aria-label" | "aria-labelledby" | "aria-level" | "aria-live" | "aria-modal" | "aria-multiline" | "aria-multiselectable" | "aria-orientation" | "aria-owns" | "aria-placeholder" | "aria-posinset" | "aria-pressed" | "aria-readonly" | "aria-relevant" | "aria-required" | "aria-roledescription" | "aria-rowcount" | "aria-rowindex" | "aria-rowspan" | "aria-selected" | "aria-setsize" | "aria-sort" | "aria-valuemax" | "aria-valuemin" | "aria-valuenow" | "aria-valuetext" | "dangerouslySetInnerHTML" | "onCopy" | "onCopyCapture" | "onCut" | "onCutCapture" | "onPaste" | "onPasteCapture" | "onCompositionEnd" | "onCompositionEndCapture" | "onCompositionStart" | "onCompositionStartCapture" | "onCompositionUpdate" | "onCompositionUpdateCapture" | "onFocus" | "onFocusCapture" | "onBlur" | "onBlurCapture" | "onChangeCapture" | "onBeforeInput" | "onBeforeInputCapture" | "onInput" | "onInputCapture" | "onResetCapture" | "onSubmit" | "onSubmitCapture" | "onInvalid" | "onInvalidCapture" | "onLoad" | "onLoadCapture" | "onError" | "onErrorCapture" | "onKeyDown" | "onKeyDownCapture" | "onKeyPress" | "onKeyPressCapture" | "onKeyUp" | "onKeyUpCapture" | "onAbort" | "onAbortCapture" | "onCanPlay" | "onCanPlayCapture" | "onCanPlayThrough" | "onCanPlayThroughCapture" | "onDurationChange" | "onDurationChangeCapture" | "onEmptied" | "onEmptiedCapture" | "onEncrypted" | "onEncryptedCapture" | "onEnded" | "onEndedCapture" | "onLoadedData" | "onLoadedDataCapture" | "onLoadedMetadata" | "onLoadedMetadataCapture" | "onLoadStart" | "onLoadStartCapture" | "onPause" | "onPauseCapture" | "onPlay" | "onPlayCapture" | "onPlaying" | "onPlayingCapture" | "onProgress" | "onProgressCapture" | "onRateChange" | "onRateChangeCapture" | "onSeeked" | "onSeekedCapture" | "onSeeking" | "onSeekingCapture" | "onStalled" | "onStalledCapture" | "onSuspend" | "onSuspendCapture" | "onTimeUpdate" | "onTimeUpdateCapture" | "onVolumeChange" | "onVolumeChangeCapture" | "onWaiting" | "onWaitingCapture" | "onAuxClick" | "onAuxClickCapture" | "onClick" | "onClickCapture" | "onContextMenu" | "onContextMenuCapture" | "onDoubleClick" | "onDoubleClickCapture" | "onDrag" | "onDragCapture" | "onDragEnd" | "onDragEndCapture" | "onDragEnter" | "onDragEnterCapture" | "onDragExit" | "onDragExitCapture" | "onDragLeave" | "onDragLeaveCapture" | "onDragOver" | "onDragOverCapture" | "onDragStart" | "onDragStartCapture" | "onDrop" | "onDropCapture" | "onMouseDown" | "onMouseDownCapture" | "onMouseEnter" | "onMouseLeave" | "onMouseMove" | "onMouseMoveCapture" | "onMouseOut" | "onMouseOutCapture" | "onMouseOver" | "onMouseOverCapture" | "onMouseUp" | "onMouseUpCapture" | "onSelect" | "onSelectCapture" | "onTouchCancel" | "onTouchCancelCapture" | "onTouchEnd" | "onTouchEndCapture" | "onTouchMove" | "onTouchMoveCapture" | "onTouchStart" | "onTouchStartCapture" | "onPointerDown" | "onPointerDownCapture" | "onPointerMove" | "onPointerMoveCapture" | "onPointerUp" | "onPointerUpCapture" | "onPointerCancel" | "onPointerCancelCapture" | "onPointerEnter" | "onPointerEnterCapture" | "onPointerLeave" | "onPointerLeaveCapture" | "onPointerOver" | "onPointerOverCapture" | "onPointerOut" | "onPointerOutCapture" | "onGotPointerCapture" | "onGotPointerCaptureCapture" | "onLostPointerCapture" | "onLostPointerCaptureCapture" | "onScroll" | "onScrollCapture" | "onWheel" | "onWheelCapture" | "onAnimationStart" | "onAnimationStartCapture" | "onAnimationEnd" | "onAnimationEndCapture" | "onAnimationIteration" | "onAnimationIterationCapture" | "onTransitionEnd" | "onTransitionEndCapture" | "prefixCls" | "allowClear" | "bordered" | "enterKeyHint" | "onPressEnter" | "addonBefore" | "addonAfter">, import("uniforms").GuaranteedProps<string> & {
        inputRef?: ((instance: Input | null) => void) | React.RefObject<Input> | null | undefined;
    }>>;
    options: {
        initialValue?: boolean | undefined;
        kind?: "leaf" | "node" | undefined;
    } | undefined;
};
export default _default;
```

</details>

<details>
<summary><code>uniforms-antd/es5/TextField.d.ts</code> after - 358B</summary>

```ts
import Input, { InputProps } from 'antd/lib/input';
import { Ref } from 'react';
import { FieldProps } from 'uniforms';
export declare type TextFieldProps = FieldProps<string, Omit<InputProps, 'onReset'>, {
    inputRef?: Ref<Input>;
}>;
declare const _default: import("uniforms").ConnectedField<TextFieldProps, string | undefined>;
export default _default;
```

</details>

---

Surprisingly, an update of TypeScript broke `connectField` inference and forced me to explicitly state `Props` in all cases. It _may_ be a breaking change type-wise but not necessarily, as it may be the case, that:
* Previously inferred type was incorrect.
* Someone already had this type parameter in place. As I checked, we actually do in a couple of projects.
Therefore, I think we can release it as a semver-minor.

### CODE REVIEW

File: packages/uniforms/src/filterDOMProps.ts [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
+export type FilterDOMPropsKeys = keyof FilterDOMProps;
 
 const registered: FilterDOMPropsKeys[] = [];
-const registeredCache = new Set<string>();
+const registeredCache = new Set<FilterDOMPropsKeys>();
 
 export const filterDOMProps = Object.assign(
   function filterDOMProps<T extends object>(props: T) {
     const filteredProps = { ...props };
     for (const prop in props) {
```
Comment (wadamek65) [inline packages/uniforms/src/filterDOMProps.ts:12]: Due to explicit typing of the cache, wouldn't this also make sense?

```suggestion
  function filterDOMProps<T extends FilterDOMPropsKeys>(props: T) {
    const filteredProps = { ...props };
    for (const prop in props) {
```

File: website/pages-parts/Tutorial/ImageField.tsx [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import React from 'react';
 import { HTMLFieldProps, connectField } from 'uniforms';
 
-type ImageProps = HTMLFieldProps<string, HTMLDivElement>;
+export type ImageFieldProps = HTMLFieldProps<string, HTMLDivElement>;
```
Comment (wadamek65) [inline website/pages-parts/Tutorial/ImageField.tsx:4]: Is the `export` necessary?

**Review by wadamek65 — APPROVED** [review]

File: packages/uniforms/src/filterDOMProps.ts (hunk 1 above)
Comment (radekmie) [reply packages/uniforms/src/filterDOMProps.ts:12]: No, as it'll require all of its properties to be required.

File: website/pages-parts/Tutorial/ImageField.tsx (hunk 2 above)
Comment (radekmie) [reply website/pages-parts/Tutorial/ImageField.tsx:4]: I just added it to match the rest.

**Review by Monteth — APPROVED** [review]

**Review by wadamek65 — APPROVED** [review]

**Pull request merged**